### PR TITLE
Experiment: make schema validation explicit.

### PR DESCRIPTION
**Pull request by sgomes**
Sorry for the large changeset in this PR, but most of it is mechanical.

**Background:** The state utils confusingly allowed for schema validation to be added to reducers in a number of ways:
- Passing a third parameter to `createReducer`
- Adding a `schema` property to a plain reducer function
- Wrapping a plain reducer function with `withSchemaValidation`

The first two, in particular, were implicit and hard to determine at compile time, leading to some strong coupling in the state utils, and requiring validation code to always be present, even if none of the reducers use it.

**Proposed solution:** This change makes schema validation explicit, by allowing for schema validation to be added to reducers via two methods:
- Using `createReducerWithValidation` (instead of `createReducer`, which loses schema support)
- Wrapping a plain reducer function with `withSchemaValidation`

This makes it possible to tree-shake effectively at compile time, should we ever drop enough schemas, and it makes state utils less coupled, by e.g. making `combineReducers` no longer depend on validation code. It may also benefit PRs like #33642 

This change should also make it easier to add types to Calypso reducers, by standardising how schemas are attached to reducers.

#### Changes proposed in this Pull Request

* Add `createReducerWithValidation`
* Remove support for the third parameter in `createReducer`
* Extract common code from `createReducer` and `createReducerWithValidation` into a utility method
* Remove support for `schema` property in plain reducer functions
* Add `package.json` with `sideEffects: false` to `state/utils` to enable tree-shaking
* Refactor all reducers to use `createReducerWithValidation` instead of `createReducer` as needed
* Refactor all reducers that were using the `schema` property to instead explicitly call `withSchemaValidation`

#### Testing instructions

I don't really have any specific testing instructions for this. The unit tests should largely cover all of this, and I added E2E tests for good measure. Also, the errors the reworked utils methods throw when unexpected usage is detected would have made most problems visible, since this all mostly happens on boot.

Take Calypso for a spin and see if anything breaks, I suppose :)

### CODE REVIEW

File: client/state/utils/create-reducer.js
Comment (sirreal) [inline client/state/utils/create-reducer.js]: Rather than a rest args, what renaming the third param and checking `typeof __deprecatedArg__ !== 'undefined'`?

```suggestion
export function createReducer( initialState, handlers, __deprecatedArg__ ) {
```

Comment (sirreal) [reply client/state/utils/create-reducer.js]: I believe `createReducer` has had limited value for some time. Does this provide anything helpful?

What if we work to remove `createReducer` completely, instead wrapping "standard" switch-based reducer function with `withSchemaValidation` as necessary?

File: client/state/utils/package.json
Comment (sirreal) [inline client/state/utils/package.json]: 😮 Is this enough to trigger webpack tree shaking for some directory tree?

**Comment by sirreal** [comment]
Full e2e passed (mobile): https://circleci.com/gh/Automattic/wp-e2e-tests-for-branches/21653

**Review by sirreal — COMMENTED** [review]
I like the direction of this and the explicitness. I'd consider going further and removing `createReducer` 🙂 

What's the status of this experiment? Initial results seem valuable and viable.

File: client/state/utils/package.json
Comment (sgomes) [reply client/state/utils/package.json]: It is indeed! Anything under that point will be tree-shaken. It blew my mind when @jsnajdr told me about it too 🤯

File: client/state/utils/create-reducer.js
Comment (sgomes) [reply client/state/utils/create-reducer.js]: > Rather than a rest args, what renaming the third param and checking `typeof __deprecatedArg__ !== 'undefined'?`

Done! Although I don't think `typeof` is necessary here, so I simplified things.

> I believe `createReducer` has had limited value for some time. Does this provide anything helpful?
>
> What if we work to remove `createReducer` completely, instead wrapping "standard" switch-based reducer function with `withSchemaValidation` as necessary?

It looks to me like `createReducer` still provides some benefits around sanity checking and serialisation, although my experience here is very limited. I'm happy to make the change here or in a follow-up PR if folks agree!

**Comment by sgomes** [comment]
> What's the status of this experiment? Initial results seem valuable and viable.

Looking at the live branch and your reaction, it's definitely positive 😄 I'm happy to call the experiment a success and merge if I get a few approvals (this touches a lot of files, so I'd appreciate having multiple pairs of eyeballs going over it 😅).

**Review by p-jackson — APPROVED** [review]
I'm generally in favour of "make a new function" over "modify behaviour of existing functions based on the presence of a magic property on one of the parameters" so definitely a fan of `withSchemaValidation` 😆

Plus smaller bundles!

A small documentation fix needed here: https://github.com/Automattic/wp-calypso/blob/master/docs/data-persistence.md#L165

**Comment by sgomes** [comment]
> I'm generally in favour of "make a new function" over "modify behaviour of existing functions based on the presence of a magic property on one of the parameters" so definitely a fan of `withSchemaValidation` 😆
> 
> Plus smaller bundles!
> 
> A small documentation fix needed here: https://github.com/Automattic/wp-calypso/blob/master/docs/data-persistence.md#L165

Thank you for the review, @p-jackson, and great catch! I've updated the data persistence docs to match the changes.

**Comment by sgomes** [comment]
E2E tests are failing for what appear to be unrelated reasons, like crashes and Gutenberg.

**Comment by sgomes** [comment]
Rebased on master.

@sirreal If you're happy with the current state of this PR, I'm happy to merge! And if there happens to be a consensus on `createReducer` being unnecessary at some point, I suggest we look at removing that in a future PR, rather than iterate on this one.

**Comment by sirreal** [comment]
> @sirreal If you're happy with the current state of this PR, I'm happy to merge!

I haven't done thorough testing or review, but I don't see any reason to hold this up and I'd be happy if you proceed 👍 

It would be good to accompany the merge with an internal post about these changes.

File: client/state/utils/create-reducer.js
Comment (jsnajdr) [reply client/state/utils/create-reducer.js]: I agree that `createReducer` is not terribly useful. Simple `switch` statement provides almost the same experience and is likely to be faster, too, as there are no lambda functions and the code is very close to C-like style.

@sgomes Does the `createReducerWithValidation` provide anything valuable over composing two simpler functions to achieve the result?
```js
const plainReducer = ( state, action ) => {
  switch ( action.type ) { ... }
};
const validatedReducer = withSchemaValidation( schema )( plainReducer );
const combinedReducer = combineReducers( { foo: validatedReducer } );
```

Comment (sgomes) [reply client/state/utils/create-reducer.js]: `createBaseReducer`, which I extracted out and is used by both `createReducer` and `createReducerWithValidation` only seems to add a sanity check for actions with invalid (falsy) types. As for `createReducerWithValidation`, it only checks to ensure that a schema is provided.

It sounds like we have a consensus on `createReducer` and `createReducerWithValidation` having outlived their usefulness 🧟‍♂️

Should I tackle that in this PR, or a separate one?

Comment (sirreal) [reply client/state/utils/create-reducer.js]: > Should I tackle that in this PR, or a separate one, @jsnajdr and @sirreal?

The work here, by itself, seems valuable. I'd also suspect this is incremental progress towards removing `createReducer` and friends. I'd vote to merge this, then continue with follow-ups.

Comment (jsnajdr) [reply client/state/utils/create-reducer.js]: I agree, let's merge and do incremental improvements later. 100+ changed files already makes a nice PR 😄

Comment (sgomes) [reply client/state/utils/create-reducer.js]: Sounds good, thank you both! 👍

**Comment by sgomes** [comment]
> It would be good to accompany the merge with an internal post about these changes.

Good idea! I'll prepare an internal post once I look into the potential `createReducer`, is that ok?

**Comment by sirreal** [comment]
> Good idea! I'll prepare an internal post once I look into the potential `createReducer`, is that ok?

I think a quick "heads up" is in order given there are a few "breaking" changes. Nothing prevents setting `reducer.schema = { … }` and it might take folks a bit to notice it's not doing what they expect.

**Comment by sirreal** [comment]
> > Good idea! I'll prepare an internal post once I look into the potential `createReducer`, is that ok?
> 
> I think a quick "heads up" is in order given there are a few "breaking" changes. Nothing prevents setting `reducer.schema = { … }` and it might take folks a bit to notice it's not doing what they expect.

Oh wait, that's an error 😅 

https://github.com/Automattic/wp-calypso/blob/67ba68be61d45c0b3607c36f5d94bb4c84116e4e/client/state/utils/reducer-utils.js#L240-L245

I still think a heads-up is in order, but not critical. Whenever you're ready 👍

**Comment by sgomes** [comment]
> I still think a heads-up is in order, but not critical. Whenever you're ready 👍

Yeah, you're right, I should take care of it sooner rather than later. I'll prepare a post with the current changes and the ones I'll be working on, as soon as I figure out the pattern I'll be using 👍

File: client/state/utils/create-reducer.js
Comment (sgomes) [reply client/state/utils/create-reducer.js]: @jsnajdr @sirreal I started looking into this more closely today, and it looks like `createReducer` adds some persistence code. Namely, if there are no specific handlers for serialization/deserialization, it wraps the reducer with the default ones:

```js
...
if ( ! handlers[ SERIALIZE ] && ! handlers[ DESERIALIZE ] ) {
  return withoutPersistence( reducer );
}
...
```

I'm not hugely familiar with this part of the codebase, so what would the consequences be of not doing so? And does this mean that we'd need to wrap every plain reducer with `withoutPersistence`, if we're trying to get rid of `createReducer`, e.g.?

```js
const isFetchingSettings = ( state = false, action ) => {
  switch ( action.type ) {
    case ACCOUNT_RECOVERY_SETTINGS_FETCH:
      return true;
    case ACCOUNT_RECOVERY_SETTINGS_FETCH_SUCCESS:
      return false;
    case ACCOUNT_RECOVERY_SETTINGS_FETCH_FAILED:
      return false;
  }
};

export default combineReducers( {
  settings,
  withoutPersistence( isFetchingSettings ),
} );
```

Comment (jsnajdr) [reply client/state/utils/create-reducer.js]: What the `withoutPersistence` wrapper does is that it makes the `SERIALIZE` action dispatch return `undefined`, meaning "we don't want to persist this part of the tree". Without it, the reducer's state will be saved to IndexedDB.

Fortunately, the `combineReducer` function automatically wraps every reducer with `withoutPersistence` before making it part of the combination. That happens only when the reducer function doesn't have a `hasCustomPersistence = true` property. That means that persistence for that reducer is already handled (i.e., one of the persistence wrapper has been applied to it) and `combineReducer` can combine it as it is.

Therefore, wrapping all reducers with `withoutPersistence` is not necessary.

Comment (sgomes) [reply client/state/utils/create-reducer.js]: That's perfect, thank you for the thorough explanation, @jsnajdr! I can forward now, then! 👍

**Pull request merged**
